lib/OpenQA/Worker.pm

```diff
@@ -14,108 +14,341 @@
         });
 
-    # register worker at startup to all webuis
-    for my $h (@{$host_settings->{HOSTS}}) {
-        # check if host`s working directory exists
-        # if caching is not enabled
+    # initialize clients to connect to the web UIs
+    my $global_settings              = $settings->global_settings;
+    my $webui_host_specific_settings = $settings->webui_host_specific_settings;
+    for my $host (@$webui_hosts) {
+        my $host_settings = $webui_host_specific_settings->{$host};
+        if (!$host_settings) {
+            die "settings for $host not correctly initialized\n";
+        }
+        my $client = $clients_by_webui_host{$host};
+        if (!$client) {
+            die "client for $host to correctly initialized\n";
+        }
+        if ($client->status ne 'new') {
+            next;
+        }
 
-        if ($worker_settings->{CACHEDIRECTORY}) {
-            $shared_cache = prepare_cache_directory($h, $worker_settings->{CACHEDIRECTORY});
+        # check if host's working directory exists if caching is not enabled
+        if ($global_settings->{CACHEDIRECTORY}) {
+            $client->cache_directory(_prepare_cache_directory($host, $global_settings->{CACHEDIRECTORY}));
         }
-        # this is being also duplicated by OpenQA::Test::Utils since 49c06362d
-        my @dirs = ($host_settings->{$h}{SHARE_DIRECTORY}, catdir($OpenQA::Utils::prjdir, 'share'));
-        ($dir) = grep { $_ && -d } @dirs;
-        unless ($dir) {
-            map { log_debug("Found possible working directory for $h: $_") if $_ } @dirs;
-            log_error("Ignoring host '$h': Working directory does not exist.");
+
+        # find working directory for host
+        # note: This is being also duplicated by OpenQA::Test::Utils since 49c06362d.
+        # FIXME: Better call this share directory?
+        my @working_dirs = ($host_settings->{SHARE_DIRECTORY}, catdir($OpenQA::Utils::prjdir, 'share'));
+        my ($working_dir) = grep { $_ && -d } @working_dirs;
+        unless ($working_dir) {
+            map { log_debug("Found possible working directory for $host: $_") if $_ } @working_dirs;
+            log_error("Ignoring host '$host': Working directory does not exist.");
             next;
         }
+        $client->working_directory($working_dir);
+        log_info("Project dir for host $host is $working_dir");
 
-        log_info("Project dir for host $h is $dir");
+        # assign other properties of the client
+        $client->worker($self);
+        $client->testpool_server($host_settings->{TESTPOOLSERVER});
+
+        # schedule registration of the web UI host
         Mojo::IOLoop->next_tick(
             sub {
-                OpenQA::Worker::Common::register_worker($h, $dir, $host_settings->{$h}{TESTPOOLSERVER}, $shared_cache);
+                $client->register();
             });
     }
 
+    return $return_code;
+}
+
+sub exec {
+    my ($self) = @_;
+
+    my $return_code = $self->init;
+
```

Comment:
would you mind saving a bit on the blank lines? Adding blank lines between each individual statement is not really helpful. I am pretty sure you can configure your editor to show it like this if you prefer ;)

Reply:
this at least you have not adressed yet :)